### PR TITLE
[0.8.x] Add Assistant API

### PR DESCRIPTION
**Pull request by gehrisandro**
https://platform.openai.com/docs/api-reference/assistants

### CODE REVIEW

**Comment by gehrisandro** [comment]
The Assistant / Threads API has a nested structured and I am currently trying to figure out, how we should access the API. Here are lists of some of the endpoints in three different variations:

1) Nested Resources
```php
$client->assistants()->create(parameters: [/* ... */]);

$client->assistants()->files()->retrieve(assistantId: 'asst_123', fileId: 'file_123');

$client->threads()->create(parameters: [/* ... */]);

$client->threads()->runs()->create(threadId: 'thread_123', parameters: [/* ... */]);

$client->threads()->runs()->submitToolOutputs(threadId: 'thread_123', runId: 'run_123', parameters: [/* ... */]);

$client->threads()->runs()->steps()->retrieve(threadId: 'thread_123', runId: 'run_123', stepId: 'step_123');

$client->threads()->runs()->steps()->list(threadId: 'thread_123', runId: 'run_123');
```

2) Nested Resources with ID parameters
```php
$client->assistants()->create(parameters: [/* ... */]);

$client->assistant(assistantId: 'asst_123')->retrieve(fileId: 'file_123');

$client->threads()->create(parameters: [/* ... */]);

$client->thread(threadId: 'thread_123')->runs()->create(parameters: [/* ... */]);

$client->thread(threadId: 'thread_123')->run(runId: 'run_123')->submitToolOutputs(parameters: [/* ... */]);

$client->thread(threadId: 'thread_123')->run(runId: 'run_123')->steps()->retrieve(stepId: 'step_123');

$client->thread(threadId: 'thread_123')->run(runId: 'run_123')->steps()->list();
```

3) Sibling Resources
```php
$client->assistants()->create(parameters: [/* ... */]);

$client->assistantFiles()->retrieve(assistantId: 'asst_123', fileId: 'file_123');

$client->threads()->create(parameters: [/* ... */]);

$client->threadRuns()->create(threadId: 'thread_123', parameters: [/* ... */]);

$client->threadRuns()->submitToolOutputs(threadId: 'thread_123', runId: 'run_123', parameters: [/* ... */]);

$client->threadRunSteps()->retrieve(threadId: 'thread_123', runId: 'run_123', stepId: 'step_123');

$client->threadRunSteps()->list(threadId: 'thread_123', runId: 'run_123');
```

@nunomaduro What do you think? ATM I would probably go for variation 1.

**Comment by obaid** [comment]
I would pick from variation 1 or 2.

**Comment by coreation** [comment]
I'd go for version 2 simply because of the "RESTful" approach of it. You define what resource you're working with as you chain the functions together. I wonder if there's a non-chaining approach so that you basically go for variation 1 but with a specific "fetch" function. I.e ...->fetchAssistantStep(threadId, runId, stepId) (or something in that nature).

Just my 2 cents, I haven't deep dived into the beta API just yet :)
In any case, thanks so much for adding the feature!

**Comment by TomasVotruba** [comment]
@gehrisandro Thanks for kicking of :+1: Already running your PR locally and trying it out :clap:

File: src/Responses/Assistant/AssistantFileListResponse.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    use HasMetaInformation;
+
+    /**
+     * @param  array<int, RetrieveJobResponse>  $data
+     */
+    private function __construct(
+        public readonly string $object,
+        public readonly array $data,
+        public readonly string $firstId,
+        public readonly string $lastId,
```
Comment (TomasVotruba) [inline src/Responses/Assistant/AssistantFileListResponse.php:34]:
```suggestion
        public readonly ?string $firstId,
        public readonly ?string $lastId,
```

These should be nullable, as for no files the null is returned.

Screenshot from 2023-11-09 16-23-17

Comment (gehrisandro) [reply src/Responses/Assistant/AssistantFileListResponse.php:34]: Makes sense. Thank you!

Currently I am implementing the "Threads" endpoints. I will push them soon.

Comment (TomasVotruba) [reply src/Responses/Assistant/AssistantFileListResponse.php]: Awesome, thank you for quick additions :clap:

**Review by nunomaduro — APPROVED** [review]

**Comment by TomasVotruba** [comment]
@gehrisandro :clap: :clap: :clap:  Thank you

**Comment by gehrisandro** [comment]
> @gehrisandro 👏 👏 👏 Thank you

You are welcome!  And thanks for supporting me!

**Comment by coreation** [comment]
Nice work @gehrisandro !

**Comment by ThibautPV** [comment]
Good job @gehrisandro 👏

Thank you !

**Comment by xitude** [comment]
Well done @gehrisandro 🎉

**Comment by heychazza** [comment]
Thank you @gehrisandro, is it possible to upload the new API somehow? I see the docs show that you can, but don't see it in the PHP wrapper

**Comment by xitude** [comment]
> Thank you @gehrisandro, is it possible to upload the new API somehow? I see the docs show that you can, but don't see it in the PHP wrapper

Can you explain by what you mean by "new API"

**Comment by heychazza** [comment]
> > Thank you @gehrisandro, is it possible to upload the new API somehow? I see the docs show that you can, but don't see it in the PHP wrapper
> 
> Can you explain by what you mean by "new API"

Hey, sorry I meant - using the files API, how would we upload a file to OpenAI?

Is it possible to upload a temp file?

**Comment by Holsta0822** [comment]
Thanks

**Pull request merged**